File: BackupUtility/Classes/NeedToCopy/NeedToCopyUpdatedOnlyChecker.cs
using System.IO;
using GenericClassLibrary.FileSystem;
using GenericClassLibrary.Logging;
using ZCopy.Interfaces;

namespace ZCopy.Classes.NeedToCopy
{
    public class NeedToCopyUpdatedOnlyChecker : INeedToCopyChecker
    {
        private readonly IFileComparer _fileComparer;
        private readonly IFileSystem _fileSystem;
        private readonly IConfirmationChecker _confirmationChecker;

        public NeedToCopyUpdatedOnlyChecker(IFileComparer fileComparer, IFileSystem fileSystem, IConfirmationChecker confirmationChecker)
        {
            _fileComparer = fileComparer;
            _confirmationChecker = confirmationChecker;
            _fileSystem = fileSystem;
        }

        public bool NeedToCopy(FolderMap baseMap, string aSource, string aTarget)
        {
            Logger.Debug($"aSource={aSource}, aTarget={aTarget}");
            Logger.Debug($"TypeOf(_fileSystem)={_fileSystem.GetType()} ");
            if (!_fileSystem.File.Exists(aTarget))
            {
                Logger.Debug($"File {aTarget} not found! Need to copy.");
                return true;
            }
            else if (_fileComparer.IsSameFile(new FileInfo(aSource), new FileInfo(aTarget)))
            {
                Logger.Debug($"File {aTarget} equal to {aSource}! Skip file!");
                return false;
            }
            return _confirmationChecker.GetConfirmation(aTarget);
        }
    }
}

File: BackupUtility/Classes/NeedToCopy/NeedToCopyUpdatedOnlyChecker.vb
Imports System.IO
Imports GenericClassLibrary.FileSystem

Public Class NeedToCopyUpdatedOnlyChecker
    Implements INeedToCopyChecker

    Private ReadOnly _fileComparer As IFileComparer
    Private ReadOnly _fileSystem As IFileSystem
    Private ReadOnly _confirmationChecker As IConfirmationChecker


    Public Sub New(fileComparer As IFileComparer, fileSystem As IFileSystem, confirmationChecker As IConfirmationChecker)
        _fileComparer = fileComparer
        _confirmationChecker = confirmationChecker
        _fileSystem = fileSystem
    End Sub

    Public Function NeedToCopy(aSource As String, aTarget As String) As Boolean Implements INeedToCopyChecker.NeedToCopy
        Dim doCopy As Boolean
        If Not _fileSystem.File.Exists(aTarget) Then
            doCopy = True
        Else
            If _confirmationChecker.GetConfirmation(aTarget) Then
                doCopy = Not _fileComparer.IsSameFile(New FileInfo(aSource), New FileInfo(aTarget))
            End If
        End If
        Return doCopy
    End Function
End Class
